File: src/server/game/config/battle_rules.go
package config

import "sync"

type BattleRules struct {
	Attackers     []string `json:"attackers"`
	TauntTypes    []string `json:"taunt_types"`
	Blocking      bool     `json:"blocking"`
	FaceAttackers []string `json:"face_attackers"`
	AttackOrder   string   `json:"attack_order"`
}

type BattleRulesConfig struct {
	BattleRules BattleRules `json:"battle_rules"`
}

var (
	battleRulesConfig    sync.Once
	BattleRulesConfigVar BattleRulesConfig
)

// GetBattleRulesConfig allows to get the game configuration linked to battle rules globally.
func GetBattleRulesConfig() BattleRulesConfig {
	return BattleRulesConfigVar
}

File: src/game/rules/battle_rules.lua
local BattleRules = {}
BattleRules.__index = BattleRules

function BattleRules:new(attackers, taunt_types, blocking, face_attackers, attack_order)
    local battleRules = setmetatable({}, BattleRules)

    battleRules.attackers = attackers or {}
    battleRules.taunt_types = taunt_types or {}
    battleRules.blocking = blocking or false
    battleRules.face_attackers = face_attackers or {}
    battleRules.attack_order = attack_order or "OneByOne"

    return battleRules
end

return BattleRules
